File: experiments/claude-wrapper/poc_wrapper.py
import pexpect
import sys
import re
import time

class Observer:
    """
    Mock Observer LLM.
    In production, this would call an actual LLM API with context.
    """
    def evaluate(self, tool_output, risk_level="unknown"):
        clean_text = re.sub(r'\x1b\[[0-9;]*m', '', tool_output)
        
        print(f"\n[Observer] Analyzing request...")
        # print(f"[Observer] Context: {clean_text.strip()[:200]}...")
        
        if "rm -rf" in clean_text:
            print("[Observer] 🛑 BLOCKED: High risk command detected.")
            return False
        
        print("[Observer] ✅ APPROVED: Safe operation.")
        return True

def run_safe_claude(initial_prompt):
    cmd = "claude" 
    print(f"[Wrapper] Starting '{cmd}'...")
    
    try:
        # Spawn the child process
        # Using a slightly larger dimensions to avoid line wrapping weirdness
        child = pexpect.spawn(cmd, encoding='utf-8', timeout=60, dimensions=(40, 120))
        child.logfile = sys.stdout
        
        # 1. Handle startup sequences
        # We assume we might see a Trust Prompt OR already be at logic
        
        while True:
            index = child.expect([
                r">",                                  # 0: Prompt ready
                r"Do you trust the files",             # 1: Trust dialog
                r"Visit .* to log in",                 # 2: Login
                r"Welcome to Claude Code",             # 3: Banner
                pexpect.EOF,
                pexpect.TIMEOUT
            ])
            
            if index == 0:
                print("\n[Wrapper] Prompt detected.")
                break # Ready to send command
                
            elif index == 1:
                print("\n[Wrapper] Trust prompt detected. Accepting...")
                child.sendline("\r") # Send Enter to accept default "Yes"
                # Loop back to wait for prompt
                continue
                
            elif index == 2:
                print("\n[Wrapper] 🚨 Auth Required! See URL above.")
                child.close()
                return
                
            elif index == 3:
                # specific banner, just wait
                print("\n[Wrapper] Welcome banner...")
                continue
                
            elif index == 4: # EOF
                print("\n[Wrapper] Unexpected EOF.")
                return
            elif index == 5: # TIMEOUT
                print("\n[Wrapper] Startup timeout.")
                return

        print("\n[Wrapper] Ready. Sending prompt...")
        child.sendline(initial_prompt)
        
        # 3. Main Event Loop
        observer = Observer()
        
        while True:
            # We look for prompts. 
            # Note: "Confirm?" prompts usually end the buffer.
            
            index = child.expect([
                r"\(y\/n\)",                   # 0
                r"\[y\/N\]",                   # 1
                r"> $",                       # 2: Main prompt returned
                pexpect.TIMEOUT,
                pexpect.EOF
            ])
            
            context = child.before
            
            if index == 0 or index == 1:
                print(f"\n[Wrapper] 🛡️  INTERCEPTION: Permission requested.")
                allowed = observer.evaluate(context)
                if allowed:
                    child.sendline("y")
                else:
                    child.sendline("n")
                    
            elif index == 2:
                # Main prompt. Check if we have output or if it's just the start.
                if len(context.strip()) > 3:
                    print("\n[Wrapper] Task seems complete (Prompt returned).")
                    break
                
            elif index == 3:
                # Timeout is not necessarily bad, might be streaming.
                # But pexpect expect loop waits for match. 
                # If streaming happens, it fills buffer but doesn't match keys.
                # We can check child.before to see progress.
                print(".", end="", flush=True)
                continue
                
            elif index == 4:
                print("\n[Wrapper] Exited.")
                break
        
        child.close()
        
    except Exception as e:
        print(f"\n[Wrapper] Error: {e}")

if __name__ == "__main__":
    prompt = "echo 'Hello Wrapper'"
    if len(sys.argv) > 1:
        prompt = " ".join(sys.argv[1:])
        
    run_safe_claude(prompt)

File: experiments/claude-wrapper/poc_wrapper.go
package main

import (
	"fmt"
	"log"
	"os"
	"os/exec"
	"regexp"
	"strings"
	"time"

	expect "github.com/Netflix/go-expect"
	"github.com/creack/pty"
)

// Observer simulates the LLM decision maker
type Observer struct{}

func (o *Observer) Evaluate(context string) bool {
	// Clean ANSI codes for analysis
	ansiRegex := regexp.MustCompile(`\x1b\[[0-9;]*m`)
	cleanText := ansiRegex.ReplaceAllString(context, "")
	cleanText = strings.TrimSpace(cleanText)

	fmt.Printf("\n[Go-Observer] Analyzing request...\n")
	// fmt.Printf("[Go-Observer] Context: %s...\n", cleanText[:min(200, len(cleanText))])

	if strings.Contains(cleanText, "rm -rf") {
		fmt.Println("[Go-Observer] 🛑 BLOCKED: High risk command detected.")
		return false
	}

	if strings.Contains(strings.ToLower(cleanText), "create") ||
		strings.Contains(strings.ToLower(cleanText), "write") ||
		strings.Contains(strings.ToLower(cleanText), "cost") {
		fmt.Println("[Go-Observer] ✅ APPROVED: Safe operation.")
		return true
	}

	fmt.Println("[Go-Observer] ⚠️ UNCERTAIN: Requesting manual review (Simulating 'Allow').")
	return true
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func RunSafeClaude(initialPrompt string) {
	fmt.Printf("[Go-Wrapper] Starting 'claude'...\n")

	// Create a new console using go-expect which manages the PTY
	c, err := expect.NewConsole(expect.WithStdout(os.Stdout), expect.WithDefaultTimeout(60*time.Second))
	if err != nil {
		log.Fatal(err)
	}
	defer c.Close()

	// Direct pty command to handle process lifecycle better if needed,
	// but go-expect can also wrap exec.Cmd.
	// Here we spawn the process manually and link it to the console's Tty.
	cmd := exec.Command("claude")
	cmd.Stdin = c.Tty()
	cmd.Stdout = c.Tty()
	cmd.Stderr = c.Tty()

	// We start the pty using creack/pty to get the pty file
	ptmx, err := pty.Start(cmd)
	if err != nil {
		log.Fatal(err)
	}
	defer ptmx.Close()

	// Link expect console to the pty master
	// Note: netflix/go-expect's usually assumes IT creates the pty or you pass the file.
	// Actually, careful here: expect.NewConsole creates a pair.
	// We want to run the command attached to that pair.
	// Let's use the simpler pattern provided by existing examples or just re-map.
	// RE-BINDING STRATEGY:
	// The best way with standard 'go-expect' is to copy streams, but we want a PTY.
	// Let's restart with the cleaner pattern: use `c.Tty()` as the slave.

	// Wait.. expect.NewConsole() creates a Master (c.File) and returns it.
	// No, it calls pty.Open() internally.

	// Let's do this:
	// We already created 'c'. cmd should use c.Tty() as its stdio.
	// But c.Tty() is the slave. Correct.

	// Re-configuring cmd to use the console's slave
	cmd = exec.Command("claude")
	slave := c.Tty()
	cmd.Stdin = slave
	cmd.Stdout = slave
	cmd.Stderr = slave

	if err := cmd.Start(); err != nil {
		log.Fatal(err)
	}

	// 1. Handle Startup sequence (Trust prompt / Banner)
	// We loop looking for the ready prompt ">"

	// go-expect's `Expect` function blocks until match.
	// We need multiple matchers.

	observer := &Observer{}

	// Startup Loop
	fmt.Println("[Go-Wrapper] Waiting for startup...")

	// Depending on state, we might hit:
	// - "Do you trust..."
	// - ">" (ready)
	// - "Login..."

	// Note: Batch expectation in go-expect is handled nicely.
	for {
		// Expect Batch
		// We look for either the Prompt or the Trust request
		val, err := c.Expect(expect.RegexpPattern(`(> $|Do you trust|Visit .* to log in|Welcome to Claude Code)`))
		if err != nil {
			log.Printf("Error expecting: %v\n", err)
			return
		}

		if strings.Contains(val, "Do you trust") {
			fmt.Println("\n[Go-Wrapper] Trust prompt detected. Accepting...")
			c.SendLine("") // Enter
			continue
		} else if strings.Contains(val, "Visit") {
			fmt.Println("\n[Go-Wrapper] 🚨 Auth Required!")
			return
		} else if strings.Contains(val, "Welcome") {
			// just wait more
			continue
		} else {
			// Matched ">"
			fmt.Println("\n[Go-Wrapper] Ready. Sending prompt...")
			break
		}
	}

	// Send the user prompt
	c.SendLine(initialPrompt)

	// Interaction Loop
	for {
		// We expect either:
		// 1. Permission request [y/N]
		// 2. The prompt ">" indicating done
		// 3. EOF/Termination

		// Regex for y/n variants
		// Go regexp: `(?i)(\[y/n\]|\(y/n\))`
		permissionPattern := `(?:\[y/N\]|\(y/n\))`
		promptPattern := `> $`

		val, err := c.Expect(expect.RegexpPattern(fmt.Sprintf("(%s|%s)", permissionPattern, promptPattern)))

		if err != nil {
			// EOF or error
			fmt.Println("\n[Go-Wrapper] Command exited or error.")
			break
		}

		// Check what we matched (idx is NOT returned by single Expect(RegexpPattern),
		// Expect returns output, match string, error.
		// NOTE: My understanding of netflix/go-expect API might be slightly off on 'idx'.
		// The `Case` API is better for multiple branches.

		// Let's check `val` (the output text matched + previous buffer).

		// The library returns the full buffer up to match + the match itself.
		// We verify the end of the string.

		// To match multiple things properly, we should use c.Expect(expect.OneOf(...)) likely?
		// But let's keep it simple: Regex OR.

		matchedText := val

		isPermission := regexp.MustCompile(permissionPattern).MatchString(matchedText)
		isPrompt := regexp.MustCompile(promptPattern).MatchString(matchedText)

		// Heuristic: If prompt is at the very end and permission pattern isn't there (or came before?)
		// Permission prompt usually ends the line.

		if isPermission {
			fmt.Printf("\n[Go-Wrapper] 🛡️  INTERCEPTION: Permission requested.\n")
			allowed := observer.Evaluate(matchedText)
			if allowed {
				c.SendLine("y")
			} else {
				c.SendLine("n")
			}
		} else if isPrompt {
			// If we matched the prompt, tasks is likely done.
			// However, sometimes permission request might be embedded? Unlikely.
			// Assuming done.
			fmt.Println("\n[Go-Wrapper] Task complete.")
			break
		}
	}

	// Cleanup
	cmd.Process.Kill()
}

func main() {
	cmd := "echo 'Hello from Go Wrapper'"
	if len(os.Args) > 1 {
		cmd = strings.Join(os.Args[1:], " ")
	}
	RunSafeClaude(cmd)
}
